### PR TITLE
Players can sign in and synchronise client settings

### PR DESCRIPTION
**Pull request by congratulatio**
This is the client portion of a player sign in feature, so we can get back the functionality of players keeping track of their game stats, levelling up, etc. Also might end up being useful for authenticating to the moderation panel on game servers.

It relies on two endpoints:
-  https://login.airmash.online/login for authentication
-  https://airmash.online/settings to save and restore settings

The code for these will be added to [airmash-refugees/airmash-backend](https://github.com/airmash-refugees/airmash-backend) soon.

### CODE REVIEW

File: src/js/UI.js [hunk 1]
```diff
@@ -567,24 +567,27 @@ UI.sayLine = function(e) {
     Players.say(e)
 };
 
-UI.escapeHTML = function(s) {
-    return (
-        ("" + s)
+UI.escapeHTML = function(s, autolink) {
```
Comment (derpsderps69) [inline src/js/UI.js]: nice catch! I also noticed some dude with a www.gaming.com or something username getting linked in the chat box. I can use the new param in a followup commit or feel free to put it in there now

File: src/js/UI.js [hunk 2]
```diff
@@ -512,14 +512,14 @@ UI.addChatLine = function(msg, text, msgType) {
     if (!ignoredPlayerIdSet[msg.id]) {
         i++;
         if (0 == msgType)
-            var o = '<div id="chat-' + i + '" class="line"><span class="playersel" data-playerid="' + msg.id + '"><span class="flag small flag-' + msg.flag + '" title="' + getFlagLabel(msg.flag) + '"></span><span class="nick">' + UI.escapeHTML(msg.name) + '</span></span><span class="text">' + UI.escapeHTML(text) + "</span></div>";
+            var o = '<div id="chat-' + i + '" class="line"><span class="playersel" data-playerid="' + msg.id + '"><span class="flag small flag-' + msg.flag + '" title="' + getFlagLabel(msg.flag) + '"></span><span class="nick">' + UI.escapeHTML(msg.name) + '</span></span><span class="text">' + UI.escapeHTML(text, true) + "</span></div>";
```
Comment (derpsderps69) [inline src/js/UI.js]: yep, I think we want ", false" in here too

**Comment by derpsderps69** [comment]
Looks good & merge at your leisure! Side thoughts:

- I wonder what Wight thinks about how this stuff should work
- As a matter of principle I almost never use these third party login services, same reason I never had an account on the original Airmash. Still a lot of people will be happy to see this returning

**Comment by derpsderps69** [comment]
Final side thought:

- Are we going to have some way to federate the database, or will this inevitably mean there is only a single login service?

File: src/js/UI.js (hunk 2 above)
Comment (congratulatio) [reply src/js/UI.js]: This is okay, `UI.escapeHTML(msg.name)` is treated the same as `UI.escapeHTML(msg.name. false)` because `undefined` evaluates false.

**Comment by congratulatio** [comment]
> Looks good & merge at your leisure! Side thoughts:
> 
> * I wonder what Wight thinks about how this stuff should work

Good point, adding @wight-airmash for review/comments.

> Final side thought:
> 
> * Are we going to have some way to federate the database, or will this inevitably mean there is only a single login service?

I've set this up with multiple Airmash clients in mind - though it's just [airmash.online](https://airmash.online) and [test.airmash.online](https://test.airmash.online) to start with, [starma.sh](https://starma.sh) could have these changes ported over and it would work the same. I think if necessary we could have some synchronisation between different login services to make sure a player logging into any such service via the same third-party account gets identified as the same player. But I reckon a single login service would be easier to manage.

**Comment by congratulatio** [comment]
A couple of points on client-server interactions: 

With these changes, when a player logs in, the login service provides them with a unique player id, which is sent to the game servers (as `LOGIN.session`) but not to other players. This is the same trust model as the original Airmash, but it does mean all server operators have the capability to impersonate players on other servers.

The login service also provides the player with a client token which is used to synchronise settings. This is not given to the game servers, because they are not responsible for keeping the settings up to date. This differs from the original Airmash, where the game servers would update the `name` and `flag` settings, and the client would only read them (from the old `/auth` endpoint).

**Comment by wight-airmash** [comment]
Very cool! 

> Are we going to have some way to federate the database, or will this inevitably mean there is only a single login service?

> But I reckon a single login service would be easier to manage.

I agree, and we will always be able to complicate the architecture, if necessary, I don't see the need to do it right now.

As for the data itself, we already have a bitter experience of data loss, so maybe we should consider how to make the data available to everyone (without violating privacy, of course), in the form of periodic dumps or otherwise.

**Comment by congratulatio** [comment]
I've added the backend server code for [login.js](https://github.com/airmash-refugees/airmash-backend/blob/master/servers/login.js) and [settings.js](https://github.com/airmash-refugees/airmash-backend/blob/master/servers/settings.js) so you can see how these fit together with the client changes.

**Comment by congratulatio** [comment]
> As for the data itself, we already have a bitter experience of data loss, so maybe we should consider how to make the data available to everyone (without violating privacy, of course), in the form of periodic dumps or otherwise.

Agreed, maybe to begin with, and ongoing for any sensitive information, a private repository could be added to [airmash-refugees](https://github.com/airmash-refugees) for synchronising such data?

**Pull request merged**